CsMerge/ProjectPackages.cs

```diff
@@ -66,22 +66,21 @@ public class ProjectPackages : IEnumerable<Package> {
       RegexOptions.CultureInvariant );
 
     public static Package PackageFromFolderName( string packageFolderName ) {
-      int indexOfVersionStart = packageFolderName.ToList().FindIndex(
-        c => {
-          int val;
-          return int.TryParse( c.ToString(), out val );
-        } );
-
-      while ( packageFolderName[indexOfVersionStart - 1] != '.' ) {
-        indexOfVersionStart++;
-      }
 
-      string id = packageFolderName.Substring( 0, indexOfVersionStart - 1 );
 
-      string versionString = packageFolderName.Substring( indexOfVersionStart );
+      var parts = packageFolderName.Split( '.' ).ToList();
 
-      var referencedPackage = new Package( id, PackageVersion.Parse( versionString ), String.Empty );
-      return referencedPackage;
+      var versionStartIndex = parts.FindIndex( c => { int val; return int.TryParse( c.ToString(), out val ); } );
+
+      if ( versionStartIndex < 0 ) {
+        throw new ArgumentException( "Unknown package folder name format.", packageFolderName );
+      } 
+        
+      var id = string.Join( ".", parts.Take( versionStartIndex ) );
+      var versionString = string.Join( ".", parts.Skip( versionStartIndex ) );
+      var packageVersion = PackageVersion.Parse( versionString );
+
+      return new Package( id, packageVersion, String.Empty );
```

Comment:
Yeah, this bit is somewhat messy :) I think this would still fail for something like Hello.1.World.1.2.3 though? It is even allowed to have something called 2.1.2.3 where the package id is "2": http://nuget.codeplex.com/wikipage?title=Package%20Id%20Specification .

The more robust approach is to look inside the nupkg file (which is a zip) for the nuspec file which specifies the id and version.